File: src/automata/lifelike.py
# pylint: disable=duplicate-code

"""Generic life-like automaton with B/S rules."""

from __future__ import annotations

from typing import Iterable, Set, Tuple

import numpy as np

from core.boundary import BoundaryMode, convolve_with_boundary

from .base import CellularAutomaton


def parse_bs(rule_str: str) -> Tuple[Set[int], Set[int]]:
    """Parse B/S rule notation like 'B3/S23' into birth/survival sets."""
    rule = rule_str.upper().replace(" ", "")
    b_part: Set[int] = set()
    s_part: Set[int] = set()

    if "B" in rule:
        try:
            start = rule.index("B") + 1
            end = rule.index("S") if "S" in rule else len(rule)
            segment = rule[start:end]
            b_part = {int(ch) for ch in segment if ch.isdigit()}
        except ValueError:
            b_part = set()

    if "S" in rule:
        try:
            start = rule.index("S") + 1
            segment = rule[start:]
            s_part = {int(ch) for ch in segment if ch.isdigit()}
        except ValueError:
            s_part = set()

    return b_part, s_part


class LifeLikeAutomaton(CellularAutomaton):
    """Generic life-like cellular automaton."""

    _KERNEL = np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])

    def __init__(
        self,
        width: int,
        height: int,
        birth: Iterable[int] | None = None,
        survival: Iterable[int] | None = None,
    ):
        self.birth = set(birth or {3})
        self.survival = set(survival or {2, 3})
        self.grid = np.zeros((height, width), dtype=int)
        super().__init__(width, height)

    def set_rules(self, birth: Iterable[int], survival: Iterable[int]) -> None:
        """Update the birth and survival rule sets."""
        self.birth = set(birth)
        self.survival = set(survival)

    def reset(self) -> None:
        """Clear the grid to an all-dead state."""
        self.grid = np.zeros((self.height, self.width), dtype=int)

    def load_pattern(self, pattern_name: str) -> None:
        """Load a named preset onto the current grid."""
        self.grid = np.zeros((self.height, self.width), dtype=int)
        if pattern_name == "Random Soup":
            random_mask = np.random.random(self.grid.shape) < 0.15
            self.grid[random_mask] = 1

    def step(self) -> None:
        """Advance the automaton by one generation."""
        bnd = BoundaryMode.from_string(self.boundary)
        neighbors = convolve_with_boundary(self.grid, self._KERNEL, bnd)
        birth_any = (
            np.logical_or.reduce([(neighbors == n) for n in self.birth])
            if self.birth
            else np.zeros_like(self.grid, dtype=bool)
        )
        survival_any = (
            np.logical_or.reduce([(neighbors == n) for n in self.survival])
            if self.survival
            else np.zeros_like(self.grid, dtype=bool)
        )
        birth = (self.grid == 0) & birth_any
        survival = (self.grid == 1) & survival_any
        self.grid = (birth | survival).astype(int)

    def get_grid(self) -> np.ndarray:
        """Return the current grid state."""
        return self.grid  # type: ignore[no-any-return]

    def handle_click(self, x: int, y: int) -> None:
        """Toggle the state of a cell at the given coordinates."""
        self.grid[y, x] = 1 - self.grid[y, x]

File: lifegrid-rs/src/automata/lifelike.rs
use rand::Rng;

use crate::core::{BoundaryMode, Grid};

use super::Automaton;

/// Generic life-like automaton with configurable B/S rule sets.
pub struct LifeLike {
    grid: Grid,
    boundary: BoundaryMode,
    birth: Vec<u8>,
    survival: Vec<u8>,
}

impl LifeLike {
    pub fn new(width: usize, height: usize, birth: Vec<u8>, survival: Vec<u8>) -> Self {
        Self {
            grid: Grid::new(width, height),
            boundary: BoundaryMode::Wrap,
            birth,
            survival,
        }
    }

    pub fn set_rules(&mut self, birth: Vec<u8>, survival: Vec<u8>) {
        self.birth = birth;
        self.survival = survival;
    }

    /// Parse a "B3/S23" style string into (birth, survival).
    pub fn parse_bs(rule: &str) -> (Vec<u8>, Vec<u8>) {
        let rule = rule.to_uppercase();
        let mut birth = Vec::new();
        let mut survival = Vec::new();
        let mut in_birth = false;
        let mut in_survival = false;
        for ch in rule.chars() {
            match ch {
                'B' => {
                    in_birth = true;
                    in_survival = false;
                }
                'S' => {
                    in_birth = false;
                    in_survival = true;
                }
                '/' => {}
                c if c.is_ascii_digit() => {
                    let d = c as u8 - b'0';
                    if in_birth {
                        birth.push(d);
                    } else if in_survival {
                        survival.push(d);
                    }
                }
                _ => {}
            }
        }
        (birth, survival)
    }
}

impl Automaton for LifeLike {
    fn name(&self) -> &'static str {
        "Custom Rules"
    }

    fn step(&mut self) {
        let neighbors = self.grid.neighbor_counts(self.boundary, 1);
        let old = self.grid.cells.clone();
        for i in 0..old.len() {
            let alive = old[i] == 1;
            let n = neighbors[i];
            self.grid.cells[i] = if alive {
                u8::from(self.survival.contains(&n))
            } else {
                u8::from(self.birth.contains(&n))
            };
        }
    }

    fn reset(&mut self) {
        self.grid.clear();
    }

    fn get_grid(&self) -> &Grid {
        &self.grid
    }

    fn get_grid_mut(&mut self) -> &mut Grid {
        &mut self.grid
    }

    fn handle_click(&mut self, x: usize, y: usize) {
        let v = self.grid.get(y, x);
        self.grid.set(y, x, 1 - v);
    }

    fn load_pattern(&mut self, _pattern: &str) {
        self.grid.clear();
        let mut rng = rand::thread_rng();
        for cell in self.grid.cells.iter_mut() {
            *cell = u8::from(rng.gen::<f32>() < 0.15);
        }
    }

    fn available_patterns(&self) -> &'static [&'static str] {
        &["Random Soup"]
    }

    fn boundary(&self) -> BoundaryMode {
        self.boundary
    }

    fn set_boundary(&mut self, mode: BoundaryMode) {
        self.boundary = mode;
    }
}
